src/lib/tls/msg_server_hello.cpp

```diff
@@ -408,4 +472,197 @@ std::vector<uint8_t> Server_Hello_Done::serialize() const
    return std::vector<uint8_t>();
    }
 
+#if defined(BOTAN_HAS_TLS_13)
+
+Server_Hello_13::Server_Hello_Tag Server_Hello_13::as_server_hello;
+Server_Hello_13::Hello_Retry_Request_Tag Server_Hello_13::as_hello_retry_request;
+
+std::variant<Hello_Retry_Request, Server_Hello_13, Server_Hello_12>
+Server_Hello_13::parse(const std::vector<uint8_t>& buf)
+   {
+   TLS_Data_Reader reader("Server_Hello_13::parse", buf);
+
+   auto data = std::make_unique<Server_Hello_Internal>(buf);
+   const auto version = data->version();
+
+   // server hello that appears to be pre-TLS 1.3, takes precedence over...
+   if(version.is_pre_tls_13())
+      { return Server_Hello_12(std::move(data)); }
+
+   // ... the TLS 1.3 "special case" aka. Hello_Retry_Request
+   if(version == Protocol_Version::TLS_V13)
+      {
+      if(data->is_hello_retry_request)
+         { return Hello_Retry_Request(std::move(data)); }
+
+      return Server_Hello_13(std::move(data));
+      }
+
+   throw TLS_Exception(Alert::PROTOCOL_VERSION,
+                       "unexpected server hello version: " + version.to_string());
+   }
+
+/**
+ * Validation that applies to both Server Hello and Hello Retry Request
+ */
+void Server_Hello_13::basic_validation() const
+   {
+   BOTAN_ASSERT_NOMSG(m_data->version() == Protocol_Version::TLS_V13);
+
+   // Note: checks that cannot be performed without contextual information
+   //       are done in the specific TLS client implementation.
+   // Note: The Supported_Version extension makes sure internally that
+   //       exactly one entry is provided.
+
+   // Note: Hello Retry Request basic validation is equivalent with the
+   //       basic validations required for Server Hello
+   //
+   // RFC 8446 4.1.4
+   //    Upon receipt of a HelloRetryRequest, the client MUST check the
+   //    legacy_version, [...], and legacy_compression_method as specified in
+   //    Section 4.1.3 and then process the extensions, starting with determining
+   //    the version using "supported_versions".
+
+   // RFC 8446 4.1.3
+   //    In TLS 1.3, [...] the legacy_version field MUST be set to 0x0303
+   if(legacy_version() != Protocol_Version::TLS_V12)
+      {
+      throw TLS_Exception(Alert::PROTOCOL_VERSION,
+                          "legacy_version '" + legacy_version().to_string() + "' is not allowed");
+      }
+
+   // RFC 8446 4.1.3
+   //    legacy_compression_method:  A single byte which MUST have the value 0.
+   if(compression_method() != 0x00)
+      {
+      throw TLS_Exception(Alert::DECODE_ERROR, "compression is not supported in TLS 1.3");
+      }
+
+   const auto& exts = extensions();
```

Comment:
This is just introduced for one usage. Simpler to write as `if(!extensions().has<Supported_Versions>())`